File: 2017/Day09.js
module.exports = function()
{
    const fs = require('fs');
    const readline = require('readline');

    const readInput = readline.createInterface({
        input: fs.createReadStream('Data/Day09.data')
    });

    readInput
    .on('line', (line) => { 
        processLine(line);
    })
    .on('close', () => {
        process.exit(0);
    });

    function processLine(line)
    {
        let garbageCount = 0;
        let count = 0;
        let index = 0;
        let length= line.length;

        function processGarbage()
        {
            if (index < length && line[index] == '<')
            {
                index++;
                while (index < length)
                {
                    let c = line[index++];
                    if (c == '!')
                        index++; // Skip next character
                    else if (c == '>') // end of garbage
                        break;
                    else
                        garbageCount++;
                }
            }
        }

        function processGroup(score)
        {
            processGarbage();

            let count = 0;

            if (index < length && line[index] == '{')
            {
                index++;
                
                count = score + processGroups(score+1);

                if (index >= length || line[index] != '}')
                    throw "Invalid input";

                index++;
            }

            processGarbage();

            return count;
        }

        function processGroups(score)
        {
            let groupCount = processGroup(score);
            
            while (index < length && line[index] == ',')
            {
                index++;
                groupCount += processGroup(score);
            }

            return groupCount;
        }

        let lineValue = processGroups(1);

        console.log("Part 1: " + lineValue + " ( 11089 )");
        console.log("Part 2: " + garbageCount + " ( 5288 )");
    }
}

File: 2017/C/Day09.c
#include <stdio.h>
#include <time.h>

static int garbageCount = 0;

static int processGroups(char** line, int score);

static void processGarbage(char** line)
{
    if (**line == '<')
    {
        (*line)++;
        while (**line)
        {
            char c = *((*line)++);
            if (c == '!')
            {
                if (**line)
                    (*line)++;
            }
            else if (c == '>')
                break;
            else
                garbageCount++;
        }
    }
}

static int processGroup(char** line, int score)
{
    processGarbage(line);

    int count = 0;
    if (**line == '{')
    {
        (*line)++;
        count = score + processGroups(line, score+1);
        if (**line != '}')
            return -1;// Error
        (*line)++;

        processGarbage(line);
    }    
    return count;
}

static int processGroups(char** line, int score)
{
    int groupCount = processGroup(line, score);
    while (**line == ',')
    {
        (*line)++;
        int value = processGroup(line, score);
        if (value < 0)
            return -1; //Error
        groupCount += value;
    }

    return groupCount;
}

void day9(void)
{
    FILE* file = fopen("Data/Day09.data", "r");

    if (file == NULL)
    {
        printf("File not found!\n");
		return;
    }
        
    static char buffer[1024 *17]; // my input is 16K
    
    if (! feof(file))
    {
        char* line = fgets(buffer, sizeof(buffer), file);
        
        int lineValue = processGroups(&line, 1);        
        if (lineValue < 0)
            printf("Invalid input\n");
        else
        {
            printf("Part 1: line value = %i ( 11089 )\n", lineValue);
            printf("Part 2: garbage count = %i ( 5288 )\n", garbageCount);
        }
    }
}

